Add config.active_record.dump_schemas.

rywall opened the pull request:
Fixes db:structure:dump when using schema_search_path and PostgreSQL extensions.

Closes #17157.

rywall committed:
Add config.active_record.dump_schemas.

Fixes db:structure:dump when using schema_search_path and PostgreSQL
extensions.

Closes #17157.

rywall commented:
@senny Here is a first pass. Feedback appreciated. :)

senny commented on activerecord/lib/active_record/core.rb:
```diff
@@ -85,6 +85,15 @@ def self.configurations
+      # Specifies which database schemas to dump when calling db:structure:dump.
+      # If :schema_search_path, it will dumps any schemas listed in schema_search_path.
+      # Use :all to always dumps all schemas regardless of the schema_search_path.
+      # A string of comma separated schemas can also be used to pass a custom list of schemas.
```
The docs should mention the default value.

rywall replied:
I was trying to model it after the documentation for schema_format (above) which doesn't specify the default in the doc. Should I change both?

senny replied:
Let's focus on this change for now. If you have time, you can update the other in a separate PR.

rywall replied:
Done.

senny commented:
@rywall looking good. Do you think you could add test-cases for this new functionality?

jimmykarily commented:
:+1:

rywall commented:
@senny I've added some tests. Anything else?

senny commented on activerecord/test/cases/tasks/postgresql_rake_test.rb:
```diff
@@ -195,21 +195,56 @@ def setup
+    end
+
+    def test_structure_dump_with_dump_schemas_string
+      ActiveRecord::Base.dump_schemas = 'test'
```
let's use an example with two schemas as well.

rywall replied:
Done.

senny commented on activerecord/test/cases/tasks/postgresql_rake_test.rb:
```diff
@@ -195,21 +195,56 @@ def setup
-      assert File.exist?(filename)
+      ActiveRecord::Tasks::DatabaseTasks.structure_dump(@configuration, @filename)
+      assert File.exist?(@filename)
     ensure
```
`ActiveRecord::Base.dump_schemas` is currently leaking outside this test. It needs to be reset:

```ruby
original_dump_schemas = ActiveRecord::Base.dump_schemas
ActiveRecord::Base.dump_schemas = "test"
# ...
ensure
  ActiveRecord::Base.dump_schemas = original_dump_schemas
```

senny commented on activerecord/test/cases/tasks/postgresql_rake_test.rb:
```diff
@@ -195,21 +195,56 @@ def setup
+
+    def test_structure_dump_with_schema_search_path_and_dump_schemas_all
+      @configuration['schema_search_path'] = 'foo,public'
+      ActiveRecord::Base.dump_schemas = :all
```
`ActiveRecord::Base.dump_schemas` is currently leaking outside this test. It needs to be reset:

```ruby
original_dump_schemas = ActiveRecord::Base.dump_schemas
ActiveRecord::Base.dump_schemas = :all
# ...
ensure
  ActiveRecord::Base.dump_schemas = original_dump_schemas
```

senny commented on activerecord/test/cases/tasks/postgresql_rake_test.rb:
```diff
@@ -195,21 +195,56 @@ def setup
+      assert File.exist?(@filename)
     ensure
-      FileUtils.rm(filename)
+      FileUtils.rm_f(@filename)
```
Why do we need `rm_f`? For a single file `rm` should be sufficient, no?

rywall replied:
I added rm_f because if the test failed and the file was not created, FileUtils.rm would raise an exception, obscuring the actual test failure. Upon reflection, the part of the test that checks if a file is created is superfluous (testing whether File.open works) so I've refactored the tests to stub out File.open and thus remove the need to cleanup.

senny commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,10 @@
+*   Add config.active_record.dump_schemas to fix db:structure:dump
```
can you put \` around "config.active_record.dump_schemas" and "db:structure:dump"?

senny commented on activerecord/test/cases/tasks/postgresql_rake_test.rb:
```diff
@@ -195,21 +195,56 @@ def setup
       }
+      @filename = "awesome-file.sql"
 
+      ActiveRecord::Base.dump_schemas = :schema_search_path
```
We better not have side-effects in the `setup` method. I'd rather modify the global state in an individual test and have that one be responsible for resetting it.

senny commented on activerecord/test/cases/tasks/postgresql_rake_test.rb:
```diff
@@ -195,21 +195,56 @@ def setup
 
+      ActiveRecord::Base.dump_schemas = :schema_search_path
       ActiveRecord::Base.stubs(:connection).returns(@connection)
+      @connection.expects(:schema_search_path)
```
I don't think we should have expectations in the `setup` method. This is something which should be related to a test-case.

senny commented:
@rywall I added a couple comments. Thank you for your work on this :yellow_heart:

rywall commented:
I've refactored the tests to avoid side effects. I think I've addressed all you other comments as well. Thanks for your feedback. The pull request is much better for it. :+1:

senny commented:
@rywall awesome! Thank you for your work :yellow_heart:

jimmykarily commented:
Thanks!

GUI commented:
Thanks, @rywall! If anyone else stumbles upon this and would find this functionality useful in Rails 3 or Rails 4, I created a gem to backport this functionality to those versions of Rails: https://github.com/GUI/activerecord-postgres-dump-schemas

njakobsen commented:
:heart: @GUI.

NikolayS commented:
Thanks for this.

Any plans to support pg_dump's `--exclude-schema` as well?

rainhead commented:
For anyone looking for a way to dump extensions along with a subset of schemas, or otherwise needing extra control over `pg_dump`, set `ActiveRecord::Tasks::DatabaseTasks.structure_dump_flags`:
```
ActiveRecord::Tasks::DatabaseTasks.structure_dump_flags = %w{--exclude-schema=schema2 --exclude-schema=schema3}
```
